### PR TITLE
Feature/lore wiki page

### PR DESCRIPTION
**Pull request by Gunvor4**
## Proposed changes

* Added a MAKE lore wiki to the internal page, so members can write wiki articles about anything lore in the MAKE universe.

## Deployment notes
* After deployment, permissions for the general MAKE group should be changed, so every member of MAKE is able to both view, add, change and delete lore articles.

## Checklist

_(If any of the points are not relevant, mark them as checked)_

- [x] Remembered to run the `makemigrations`, `makemessages` and `compilemessages` management commands and committed any changes that should be included in this PR
- [x] Created tests that fail without the changes, if relevant/possible
- [x] Manually tested that the website UI works as intended with different device layouts
  - _(Most common is to test with typical screen sizes for mobile (320-425 px), tablet (768 px) and desktop (1024+ px), which can easily be done with your browser's dev tools)_
- [x] Manually tested that everything works as intended when logged in as different users locally
  - _(This can be e.g. anonymous users (i.e. not being logged in), "normal" non-member users, members of different committees, and superusers)_
- [x] Made sure that your code conforms to [the code style guides](https://github.com/MAKENTNU/web/blob/main/CONTRIBUTING.md#code-style-guides)
  - _(It's not intended that you read through this whole document, but that you get yourself an overview over its contents, and that you use it as a reference guide / checklist while taking a second look at your code before opening a pull request)_
- [x] Attempted to minimize the number of [common code smells](https://github.com/MAKENTNU/web/blob/main/CONTRIBUTING.md#code-review-guideline-code-smells)
  - _(See the comment for the previous checkbox)_
- [x] Added sufficient documentation - e.g. as comments, docstrings or in the [README](https://github.com/MAKENTNU/web/blob/main/README.md), if suitable
- [x] Added your changes to the ["Unreleased" section of the changelog](https://github.com/MAKENTNU/web/blob/main/CHANGELOG.md#unreleased) - mainly the changes that are of particular interest to users and/or developers, if any
- [x] Added a "Deployment notes" section above, if anything out of the ordinary should be done when deploying these changes to the server
- [x] Structured your commits reasonably
  - _(Some tips on this can be found in ["Write Better Commits, Build Better Projects"](https://github.blog/2022-06-30-write-better-commits-build-better-projects/) and in [the code style guides](https://github.com/MAKENTNU/web/blob/main/CONTRIBUTING.md#commit-message))_
  - _(Restructuring commits after they've already been committed, can be done using e.g. interactive rebase; see [the Git docs on rewriting history](https://git-scm.com/book/en/v2/Git-Tools-Rewriting-History))_

### CODE REVIEW

**Comment by Gunvor4** [comment]
Fullskjermvisning_1
Fullskjermvisning_2
Mobilvisning_1
Mobilvisning_2

**Comment by elisakiv** [comment]
It looks great, just few comments:

It seems that when the title of an article is one very long word with no spaces, it behaves strangely on mobile. The name does not wrap when it becomes too long so the title continues on one line, past the length of the header.

**Comment by elisakiv** [comment]
If a users adds an article with a name only consiting of special character, the page crashes.

**Comment by elisakiv** [comment]
When editing an article with a picture, the button and checkbox for removing the picture behaves a little strangely. The checkbox is not level with the picure name, and the "Clear" button does not fully behave like a button, mainly because the cursor does not become a hand,  the user does not receive any feedback that the button was clicked, and it does not look like a button either, it looks like the other text on the page.

**Comment by elisakiv** [comment]
If the article has a very long name, the title in the menu overlaps with the text of the article that the user has open

**Review by elisakiv — APPROVED** [review]

File: src/internal/static/internal/css/make_lore.css [hunk 1]
```diff
@@ -0,0 +1,173 @@
...
+    max-width: 100vw;
+}
+
+#lore-wiki-list {
+    float: left;
+    width: 20%;
+    padding: 3rem 0 40px 40px;
+    position: relative;
+    overflow-wrap: break-word;
+    hyphens: auto;
```
Comment (ddabble) [inline src/internal/static/internal/css/make_lore.css:13]: Use the `word-breaking` and `hyphenated` classes (defined in `style.css`) in your HTML code instead of using these properties 🙂 (This also applies to some other selectors below.)

File: src/internal/templates/internal/lore_wiki.html [hunk 2]
```diff
@@ -0,0 +1,72 @@
...
+{% endblock title %}
+
+{% block header_links %}
+    <link rel="stylesheet" href="{% static 'internal/css/make_lore.css' %}">
+    <link rel="stylesheet" href="{% static 'web/css/header.css' %}"/>
+    <script defer src="{% static 'web/js/header.js' %}"></script>
+{% endblock header_links %}
+
+{% block body %}
+    <div id="lore-page">
```
Comment (ddabble) [inline src/internal/templates/internal/lore_wiki.html:17]: Use a `<div>` with the classes `ui text container` as the outermost element, like pretty much all of the other templates do 🙂

File: src/internal/templates/internal/lore_wiki.html [hunk 3]
```diff
@@ -0,0 +1,72 @@
...
+                    <a href="{% url 'add_lore' %}">
+                        <i class="green plus icon"></i>
+                    </a>
+                {% endif %}
+            </h2>
+            {% include 'internal/includes/lore_articles_list.html' %}
+        </div>
+
+        {# Checkbox belongs to hamburger menu. Checkbox status decides if hamburger menu will be shown or not #}
+        <input type="checkbox" id="burger-input">
```
Comment (ddabble) [inline src/internal/templates/internal/lore_wiki.html:30]: Neat trick! 💯

File: src/internal/templates/internal/lore_wiki.html [hunk 4]
```diff
@@ -0,0 +1,72 @@
...
+
+{% block title %}
+    {% translate "The wiki of MAKE lore" %}
+{% endblock title %}
+
+{% block header_links %}
+    <link rel="stylesheet" href="{% static 'internal/css/make_lore.css' %}">
+    <link rel="stylesheet" href="{% static 'web/css/header.css' %}"/>
+    <script defer src="{% static 'web/js/header.js' %}"></script>
+{% endblock header_links %}
```
Comment (ddabble) [inline src/internal/templates/internal/lore_wiki.html:14]: I think you mistakenly overrode `header_links` instead of `extra_head`, which would explain why you linked `header.css`  and `header.js` here 😅

```suggestion
{% block extra_head %}
    <link rel="stylesheet" href="{% static 'internal/css/make_lore.css' %}">
{% endblock extra_head %}
```

File: src/internal/templates/internal/lore_wiki.html [hunk 5]
```diff
@@ -0,0 +1,72 @@
...
+            {% endif %}
+        </h2>
+        <div id="lore-burger-menu">
+            {% include 'internal/includes/lore_articles_list.html' %}
+        </div>
+
+        <div id="lore-article">
+            {% if show_article %}
+                <h1 class="lore-heading">{{ lore_article.title|capfirst }}</h1>
+                {% if perms.internal.delete_lore and lore_article.title != "Dev" %}
```
Comment (ddabble) [inline src/internal/templates/internal/lore_wiki.html:46]: `and lore_article.title != "Dev"` lol

File: src/internal/templates/internal/lore_wiki.html [hunk 6]
```diff
@@ -0,0 +1,72 @@
...
+                    </a>
+                {% endif %}
+                <div class="lore-container">
+                    {% if lore_article.image %}
+                        <img class="lore-img" src="{{ lore_article.image.url }}" alt="{{ lore_article.title }}">
+                    {% endif %}
+                    <p>{{ lore_article.text|linebreaks }}</p>
+                </div>
+            {% else %}
+                <h1>{% translate "Welcome to the wiki of MAKE lore" %}</h1>
```
Comment (ddabble) [inline src/internal/templates/internal/lore_wiki.html:65]: `<h1>` should not be used here when the headings above are `<h2>`, as this illogical heading hierarchy makes the page less accessible for users with screen readers. Instead, make sure that all the heading levels (1-6) are in logical order, and then style them using either CSS rules or [the `ui <size> header` classes](https://fomantic-ui.com/elements/header.html#content-headers), where `<size>` ranges from `tiny` to `huge` (see the linked docs).

File: src/internal/static/internal/css/make_lore.css
Comment (ddabble) [inline src/internal/static/internal/css/make_lore.css:1]: Some of the code related to the burger icon could maybe look like it was copied from elsewhere on the internet; this is completely fine, of course, but if that's the case, you should add a comment saying where you found it / what you based it on (e.g. `/* Burger code based on <link> */`) 🤠

**Review by ddabble — CHANGES_REQUESTED** [review]
Just a few comments 😅

I guess I could comment more on the CSS, but I can potentially wait with that until after you've made some of the changes I suggested below - if you decide to 🙂 (Otherwise, I must admit that I personally don't find CSS code in general to be particularly important to keep polished :))

Finally, it would also be very nice if you could add some (simple) tests for the new code, so that we can feel more secure that it keeps working the way we think it does - both now and after future changes to the codebase 🙂

File: src/internal/models.py [hunk 7]
```diff
@@ -314,3 +316,34 @@ class Meta:
...
+
+
+def internal_subclass_directory_path(instance: 'Lore', filename: str):
+    model_name = instance._meta.model_name
+    return f"internal/{model_name}s/{filename}"
+
+
+class Lore(models.Model):
+    title = models.CharField(max_length=50, verbose_name=_("title"), null=False)
+    slug = models.SlugField()
```
[the remark this answers is not in the document]
Comment (Gunvor4) [reply src/internal/models.py:328]: Because this is a wiki, and wikis usually use slugs, not pk's. Yes, it does add complexity to the code, but I still think a slug looks better than a pk in this case.

File: src/internal/views.py [hunk 8]
```diff
@@ -332,3 +333,100 @@ def has_permission(self):
...
+
+class LoreDetailView(DetailView):
+    template_name = 'internal/lore_wiki.html'
+    model = Lore
+    context_object_name = 'lore_article'
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context['show_article'] = True
+        lore_topics = Lore.objects.all().order_by('title')
```
[the remark this answers is not in the document]
Comment (Gunvor4) [reply src/internal/views.py:358]: This was done to distinguish between the article being shown in the detail view, and the topics being shown in the menu (belonging to the list view, or to the extra context in the detail view). It made it easier to use the same template both in the list view and in the detail view.

File: src/web/static/web/css/style.css [hunk 9]
```diff
@@ -209,6 +209,16 @@ span.no-wrap {
...
+input[type=checkbox]#image-clear_id {
+    display: inline;
+    position: relative;
+    top: 7px;
+    margin-left: 10px;
+}
+
+label[for=image-clear_id] {
+    display: inline !important;
+}
```
[the remark this answers is not in the document]
Comment (Gunvor4) [reply src/web/static/web/css/style.css:221]: This was done because the CSS code in question applies to the base form, which uses the style.css file because it is used different places across the webpage. I added it because when the form is used by an update view, and the model instance has an image attached to it, the form is displayed in a very unlogical way. The checkbox for removing the image is displayed after the file name of the image, while "remove" is on the next line.
Since I now switched from having an image field on my model, to simply using RichTextUploadingField instead of Text field, this is no longer an issue in the lore update form, but it might still be an issue in other update forms.

File: src/internal/models.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ddabble) [reply src/internal/models.py:328]: I wouldn't say that other wikis using slugs is really an argument for why we should use them; that being said, the underlying reason is usually either or both of:
1. It prevents enumerating resources, if that's important to the website owners, for whatever reason - usually for security or business reasons
2. It makes it somewhat easy to see at a glance the title of the resource - as long as the URL is not too long or cluttered, of course
   - This is often made redundant, however, as when you paste a URL into e.g. Slack, Facebook or most other social media / messaging services, a small preview of the website is usually shown, which takes info (incl. a webpage title) from the Open Graph data defined on each specific page (see e.g. [how articles and events insert this data](https://github.com/MAKENTNU/web/blob/main/src/news/templates/news/news_base_detail.html#L9-L13))

I guess the latter reason is the one most relevant to us, and it is absolutely a good reason, but I personally think it has more downsides than upsides. The only advantage that I can see is the one mentioned above, but the disadvantages weigh heavier, I think:
1. The main one is that it introduces complexity that adds a maintenance cost
2. As the code is right now, if a user changes the title of a lore article, existing URLs to that article will no longer be valid
   - This can be fixed by making sure that the slug is never changed after a lore article is created, but then you might lose the original advantage of having slugs in the first place, if the title and/or content is changed to no longer be reasonably summarized by the original title/slug

So I'd recommend that you keep the code as simple as it can be (without losing the functionality that you want, of course), but ultimately that decision is for you guys to take 🙂

File: src/internal/views.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ddabble) [reply src/internal/views.py:358]: In that case, I think better names for the variables could be to describe them pretty much like you did just now, so the shown lore object could be named e.g. `shown_lore_article`, and the other lore objects could be `all_lore_articles`. As the code is now, it's very easy for new readers to think that lore topics and lore articles are two distinct types of objects (which I initially thought), and so I'd urge you to change it 🙂

File: src/web/static/web/css/style.css (hunk 9 above)
[the remark this answers is not in the document]
Comment (ddabble) [reply src/web/static/web/css/style.css:221]: Ah right, while partly skimming through the code, I thought this was specific to the lore pages 😅 In that case, the issue can be more localized to [the `semantic_file.html` template](https://github.com/MAKENTNU/web/blob/a90ac79f5756721c9a3864658a87fa62633dbc6c/src/web/templates/web/forms/widgets/semantic_file.html), and so this code should be moved to [`semantic_file.css`](https://github.com/MAKENTNU/web/blob/a90ac79f5756721c9a3864658a87fa62633dbc6c/src/web/static/web/css/forms/widgets/semantic_file.css).

[the remark this answers is not in the document]
Comment (ddabble) [reply src/web/static/web/css/style.css:221]: Oh wait, the lore form doesn't use the Fomantic UI widgets, as far as I can see :o Also, I don't think `semantic_file.html` currently supports clearing a submitted value, so I guess it's easiest not to worry about those widgets, and instead move the CSS code to a new CSS file that can be linked through [a `Media` class](https://docs.djangoproject.com/en/4.2/topics/forms/media/) added to all forms that use Django's vanilla file widget (which currently seems like only applies to `LoreForm`).

Ideally, `semantic_file.html` should be changed to support clearing a submitted value, but that might be out of scope of this PR.. Also I've already done it in a local branch, which I'll push sometime later ™️ 🤠

File: src/internal/views.py [hunk 10]
```diff
@@ -332,3 +333,100 @@ def has_permission(self):
...
+        context['lore_topics'] = lore_topics
+        return context
```
[the remark this answers is not in the document]
Comment (Gunvor4) [reply src/internal/views.py:360]: When I switch to using extra_context instead of overriding the get_context_data method, the list of lore articles added via the extra context does not update when a new lore article is added. So I am keeping this the way it is.

File: src/internal/static/internal/css/make_lore.css (hunk 1 above)
Comment (Gunvor4) [reply src/internal/static/internal/css/make_lore.css:13]: This works for the title and the menu, but for some reason not for the <p> element displaying the lore content. So for that particular field, I am keeping overflow-wrap and hyphens in my local css-file.

File: src/internal/templates/internal/lore_wiki.html (hunk 2 above)
Comment (Gunvor4) [reply src/internal/templates/internal/lore_wiki.html:17]: This messes up the entire look of the page, and gives me too large margins.

File: src/internal/templates/internal/lore_wiki.html (hunk 6 above)
Comment (Gunvor4) [reply src/internal/templates/internal/lore_wiki.html:65]: I don't really understand the problem here. This HTML file is used as a template by two different views - the general list view that users land on when they first access the lore page, and the detail view that shows a particular lore article. I use The <h1> heading commented on here, is the general heading used by the list view. When a lore article is selected, the heading for the article is also displayed in an <h1> tag.

File: src/web/static/web/css/style.css (hunk 9 above)
[the remark this answers is not in the document]
Comment (Gunvor4) [reply src/web/static/web/css/style.css:221]: But then everyone needing a form would have to include the media class with the correct css-file in their forms, which they will probably not do if they don't know about this separate CSS-file. Can't I just move this code to a separate file, and include this new file in the generic_form.html template instead?

File: src/internal/views.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ddabble) [reply src/internal/views.py]: Ah, true, I forgot that `'all_lore_topics'` is a queryset, which don't get refreshed between each request when assigned to a class field - as I mentioned in [this comment on another PR](https://github.com/MAKENTNU/web/pull/669#discussion_r1403798854) 😅

Context variables that are supposed to be constant, however, can preferably be listed in `extra_context`, I think, as it makes the view code ever so slightly clearer :) So:
```python
extra_context = {
    'show_article': True,
}
```

File: src/web/static/web/css/style.css (hunk 9 above)
[the remark this answers is not in the document]
Comment (ddabble) [reply src/web/static/web/css/style.css]: All files linked through widgets' and forms' `Media` class are automatically linked when inserting the form's `media` property into the form template (see [how the `media` property works](https://docs.djangoproject.com/en/stable/topics/forms/media/#media-on-forms)) - which should always be done when working with a form. When inheriting from `CustomFieldsetFormMixin` or otherwise using the `generic_form.html` template, [`form.media` is automatically inserted](https://github.com/MAKENTNU/web/blob/a90ac79f5756721c9a3864658a87fa62633dbc6c/src/web/templates/web/generic_form.html#L8) 🙂

However, seeing as you removed the `image` field, nevermind this 😅

File: src/internal/templates/internal/lore_wiki.html (hunk 2 above)
Comment (ddabble) [reply src/internal/templates/internal/lore_wiki.html]: Larger margins than on the other pages using `ui text container`? If they're the same, then that is very much the point, as [lines should not be too wide (seems like around 70-80 characters is the convention) for accessibility reasons](https://www.w3.org/WAI/tutorials/page-structure/styling/#line-length) 🙂 _(Though kind of ironic that the linked page does not have a `max-width` set for its text 😅)_

So for readability and a consistent layout across all the pages of the website, I'd recommend starting with `ui text container` as a baseline, and using CSS properties like `position: absolute/fixed/sticky` to place elements outside of the margins, if that's wanted - like [the "Topics" sidebar you posted a screenshot of above](https://github.com/MAKENTNU/web/pull/675#issuecomment-1711822749).

File: src/internal/templates/internal/lore_wiki.html (hunk 6 above)
Comment (ddabble) [reply src/internal/templates/internal/lore_wiki.html]: Ah yes, I didn't run the code before reviewing, so I might have spoken too soon here 🤔 I don't have a huge amount of experience with web accessibility, but judging by the page's accessibility tree view (opened in Chrome through the button in the screenshot below), the hierarchy looks okay, I think 🙂
> Screenshot of button for opening Google Chrome's accessibility tree view

Though I guess it could be slightly improved by placing the sidebar and mobile menu elements inside a [`<nav>` element](https://developer.mozilla.org/en-US/docs/Web/HTML/Element/nav), as their main purpose is to provide navigation between the lore articles, and not as a part of the page's main contents 🤔 In that case, an `aria-label`/`aria-labelledby` attribute should be added to the `<nav>` element, e.g. `aria-label="{% translate "Lore articles" %}"`.

File: src/internal/admin.py [hunk 11]
```diff
@@ -93,7 +93,13 @@ class QuoteAdmin(DefaultAdminWidgetsMixin, UserSearchFieldsMixin, admin.ModelAdm
     autocomplete_fields = ('author',)
 
 
+class LoreAdmin(admin.ModelAdmin):
+    ordering = ('title',)
```
Comment (ddabble) [inline src/internal/admin.py:97]: I can imagine that it could be useful to also have a column with `slug` 🤔 What do you think?
```suggestion
class LoreAdmin(admin.ModelAdmin):
    list_display = ('title', 'slug')
    ordering = ('title',)
```

File: src/internal/models.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ddabble) [reply src/internal/models.py]: In any case, as the code is now, point 2 above (the bottom one) should be addressed, in my opinion. The alternatives I can think of are:
1. Disallowing changing the title after creation
2. Not updating the slug after changing the title
3. Keeping track of all previous slug values and redirecting from the old ones to the newest one
   * This would also require not allowing any lore articles to be renamed so that their slugs collide with any previous slugs of other articles

Any thoughts on this? 🙂

**Review by ddabble — CHANGES_REQUESTED** [review]
Just doing a second round after the most recent commits 🙂

I just remembered that I might not have taught you anything about writing tests..? If so, just send a DM to me (or maybe @elisakiv..?) if you'd like some pointers :)

In any case, I wanted to reiterate that it would be great if you could add some (simple) tests for the lore pages, which would both help us feel more secure that the code behaves as we think it does - especially in edge cases - and help future developers understand the intended (main) functionality of the code by showing some typical scenarios in some of the tests - as tests often do. Examples include:
* That a MAKE member (i.e. a user with some of the lore permissions) is able to view/add/change(/delete?) lore pages (through GET and/or POST requests), but an unprivileged user is not
* That the slug field has the expected value after submitting various titles
* That colliding slugs/titles are correctly identified
* That the slug/title field cannot be empty
* Anything else that you think would be useful to have tests for 🙂

File: src/internal/models.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ddabble) [reply src/internal/models.py]: _(Reposted this comment, as the previous one somehow appeared both here and above my most recent review; probably a bug..? 🤷)_

In any case, as the code is now, point 2 above (the bottom one) should be addressed, in my opinion. The alternatives I can think of are:
1. Disallowing changing the title after creation
2. Not updating the slug after changing the title
3. Keeping track of all previous slug values and redirecting from the old ones to the newest one
   * This would also require not allowing any lore articles to be renamed so that their slugs collide with any previous slugs of other articles

Any thoughts on this or additional ideas? 🙂

**Comment by ddabble** [comment]
Btw, regarding the deployment notes in the PR description, it might be wise to not let all members of MAKE *delete* lore articles, and instead only grant that permission to e.g. the board members - similar to the group permissions for the internal quotes currently on the website 🤔 What do you think?

File: src/internal/views.py [hunk 12]
```diff
@@ -332,3 +333,74 @@ def has_permission(self):
...
+class LoreDetailView(DetailView):
+    model = Lore
+    template_name = 'internal/lore/lore_list.html'
+    context_object_name = 'shown_lore_article'
+
+    def get_context_data(self, **kwargs):
+        context = super().get_context_data(**kwargs)
+        context['show_article'] = True
+        context['all_lore_articles'] = Lore.objects.order_by('title')
+        return context
```
[the remark this answers is not in the document]
Comment (Gunvor4) [reply src/internal/views.py:357]: I saw the previous comment, but frankly I don't understand the advantages of doing it the way you propose, so it's a bit frustrating to have to change something that works when I don't understand why I am changing it. I have now changed it, but I'd appreciate an explanation to why the second version of this code is better than the first.

File: src/internal/forms.py [hunk 13]
```diff
@@ -176,3 +178,35 @@ class Meta:
...
+        widgets = {
+            'text': CKEditorWidget()
+        }
+
+    def clean(self):
+        cleaned_data = self.cleaned_data
+        title = cleaned_data.get('title')
+        title = title.replace('ø', 'o')
+        title = title.replace('æ', 'ae')
+        slug = slugify(title)
```
[the remark this answers is not in the document]
Comment (Gunvor4) [reply src/internal/forms.py:196]: I agree, much better solution. The only reason I did that very ad-hoc replace-solution in the first place was because I didn't know it was possible to fix this by simply writing `allow_unicode=True`

File: src/internal/forms.py [hunk 14]
```diff
@@ -176,3 +178,35 @@ class Meta:
...
+        try:
+            article = self.instance
+        except NameError:
+            article = None
+        if article:
+            if (article.slug != slug) and (Lore.objects.filter(slug=slug).exists()):
```
[the remark this answers is not in the document]
Comment (Gunvor4) [reply src/internal/forms.py:202]: They improve readability to me, but if I'm the only one who likes parentheses, I'll remove them.

File: src/internal/models.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (Gunvor4) [reply src/internal/models.py]: I now created an additional model storing changes in slugs, so URLs consisting of slugs that are not used any more, will still be valid. If two different lore articles at different points in time have had the same slug, the user will be redirected to the one that appear first in the HistoricSlug model.

File: src/internal/views.py (hunk 12 above)
[the remark this answers is not in the document]
Comment (ddabble) [reply src/internal/views.py:357]: Sure, that's fair 🙂 So in this case, I guess the explanation has two parts (I'm guessing you were mostly interested in the first half of the second one..?):
1. "Why move context variables from `get_context_data()` to `extra_context`?"

   This is generally because there's less "noise" when reading it (compare e.g. `extra_context = {...}` to `def get_context_data(self, **kwargs): return {...}`), and so it's preferable whenever possible - which is the case when the values are constant (e.g. `True`), but not when they're variable/dynamic (e.g. `Lore.objects.order_by('title')`).

   In this case, however, one of the values is constant, while the other is variable, and so having _both_ `extra_context` and `get_context_data()` can make the code even noisier, but it can also be worth keeping the constant and variable variables separated, for the sake of organization - it's hard to say that either approach will always be the best choice 🤷 But since `LoreListView` already has an `extra_context` with `'show_article'`, I think it's best to have its "sister view", `LoreDetailView`, match that, if that makes sense :)
2. "Why structure the code inside `get_context_data()` the way I suggested?"

   This is purely to follow a principle of designing your code to be as future-proof as practically possible (i.e. as long as it doesn't take substantially more effort compared to a "naive" design). To illustrate why I think my suggestion adheres more closely to that principle, take the following example of code that could potentially be written in the future:
   * Writing a test for how the view and/or template behaves when some of the context variables are different (for whatever reason):
     ```python
     lore_detail_view_instance.get_context_data(show_article=False, all_lore_articles=None)
     ```
     This works with the code I suggested, but with the original code, `'show_article'` would have been overwritten (or "forced") to always have the value `True` regardless of the kwargs provided - and likewise for `'all_lore_articles'` being overwritten to equal `Lore.objects.order_by('title')`.

   That's what I meant when I said "[...] always let other code calling a context data method to be able to override the values set in the method, using `kwargs`" in the other comment.

   Also, as an added bonus, PyCharm seems to have a feature of providing code suggestions in templates for the context variables whose names are directly mentioned in the dict literal (i.e. `{...}`) returned by `get_context_data()` - but only when we write `return {...}` and not e.g. `context = {...}` followed by `return context`. (This is most likely because it's easier to parse one dict literal you know the location of (i.e. directly following the method's sole `return` statement), than to somehow analyze which context variables are inserted into `context` and then later returned 🤷)

Did that explanation make sense? If not, which details was I e.g. too vague about? 🙂

File: src/internal/models.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (ddabble) [reply src/internal/models.py]: > If two different lore articles at different points in time have had the same slug, the user will be redirected to the one that appear first in the HistoricSlug model.

Hm, this still has the potential of causing confusion when a URL can change which lore article it points to, if a future lore article happens to use the same slug as an existing historical one 🤔 Though it does seems like it would be better than the current solution 🙂

---

Now, you didn't comment on the first part of [my second comment above](#discussion_r1391555887), but I'm assuming that the practical reason you want slugs in the URLs in the first place (apart from the aesthetics of the URLs), is to make it somewhat easy to read the title of the page from the URL - which in turn might act as a way of communicating to users that they're URLs to a kind of wiki (since no parts of the lore URLs contain the word "wiki"). Is this correct?

If so, before going any further with the solution you've already created, I recently had an idea that pretty much eliminates all the problems with using slugs that I've previously mentioned, _and_ should drastically reduce the amount of code! Simply let the slugs be a "decorative" part of the URLs, but have the identifying part of the URLs be the lore articles' PK! This could e.g. look like `https://i.makentnu.no/lore/64/the-title-of-a-lore-article/` (or, alternatively, `https://i.makentnu.no/lore/the-title-of-a-lore-article/64/` - though I personally prefer the first suggestion). Furthermore, lore URLs without the slug or with a wrong/outdated slug can be redirected to the URL with the correct, up-to-date slug. This can be implemented by doing the following:
1. Replace `<slug:slug>` with `<int:pk>` in the `path()` that includes `specific_lore_urlpatterns`
2. Duplicate the `path()` of `LoreDetailView` and add `<str:slug>/` to the path of the new one (so that both paths reference `LoreDetailView`), and placing that path **after** the `change/` and `delete/` paths - since both of those strings would match `<str:slug>/` and therefore essentially make it impossible to reach those two paths
   * Note that I wrote `str:slug`, not `slug:slug`, because we want the view to be able to "correct" the URLs no matter what the last part of the URL is
3. Make sure that the code inside `LoreDetailView` ignores `slug` completely when it queries the database for the lore object, since only the PK will be used to identify lore articles
4. Inside `LoreDetailView`, override its [`get()`](https://docs.djangoproject.com/en/stable/ref/class-based-views/generic-display/#django.views.generic.detail.BaseDetailView.get) method, check the URL's `slug` against the lore object's `slug`, and if they differ, return a redirect response using the correct slug
5. Remove `Lore`'s `slug` field and replace it with a `slug` `@property`, since the slug is not needed anymore for querying, but only for display; also remove overriding `save()` (as mentioned in [this comment](https://github.com/MAKENTNU/web/pull/675/files#r1425896596)), and add `get_absolute_url()` (as mentioned in [this comment](https://github.com/MAKENTNU/web/pull/675/files#r1426077750))
6. This should result in roughly the following code:
   ```python
   # internal/urls.py:

   specific_lore_urlpatterns = [
       path("", views.LoreDetailView.as_view(), name='lore_detail'),  # Redirects to `lore_detail` below
       path("change/", views.LoreUpdateView.as_view(), name='lore_update'),
       path("delete/", views.LoreDeleteView.as_view(), name='lore_delete'),
       # Must be placed at the bottom of this list to not "hide" the paths above
       # (The slug is purely decorative; the PK is the identifying part of the path)
       path("<str:slug>/", views.LoreDetailView.as_view(), name='lore_detail'),
   ]
   lore_urlpatterns = [
       path("", views.LoreListView.as_view(), name='lore_list'),
       path("add/", views.LoreCreateView.as_view(), name='lore_create'),
       path("<int:pk>/", include(specific_lore_urlpatterns)),
   ]

   # internal/views.py:

   class LoreDetailView(DetailView):
       model = Lore
       slug_url_kwarg = None  # Ignore the path's `slug` parameter
       ...  # template_name and context_object_name
   
       def get(self, request, *args, **kwargs):
           if slug := kwargs.get('slug'):
               lore_article = self.get_object()
               if slug != lore_article.slug:
                   return HttpResponseRedirect(lore_article.get_absolute_url())
   
           return super().get(request, *args, **kwargs)
   
       def get_context_data(self, **kwargs):
           ...

   # internal/models.py:

   class Lore(models.Model):
       ...  # fields (with `slug` removed), the `Meta` class, and methods (with `save()` removed)
   
       def get_absolute_url(self):
           return reverse_internal('lore_detail', self.pk, self.slug)
   
       @property
       def slug(self):
           return slugify(self.title, allow_unicode=True)
   ```

What do you think? I haven't tested the code above, so please let me know if you encounter any bugs with it that you're unsure about how to solve :)

If you agree with implementing this idea, I can absolutely understand if you're also hesitant to scrap all the work you've put into your original idea - that's never particularly fun 🙃 Though (and I'm saying this with empathy, as I've been there myself) hopefully this can help remind you in the future to consider discussing the thing you're creating before spending time creating it - for example with the people you plan on inviting to review your work 😊

File: src/internal/models.py [hunk 15]
```diff
@@ -314,3 +314,24 @@ class Meta:
...
+
+    def save(self, *args, **kwargs):
+        title = self.title
+        self.title = capfirst(title)
+        title = title.replace('ø', 'o')
+        title = title.replace('æ', 'ae')
+        self.slug = slugify(title)
+        super().save(*args, **kwargs)
+
+    class Meta:
```
Comment (ddabble) [inline src/internal/models.py:335]: Ah btw, the `Meta` class should be placed before `__str__()`, as per [the code style guide](https://github.com/MAKENTNU/web/blob/main/CONTRIBUTING.md#field-and-method-order) :)
